schain/schainwriter_test.go

```diff
@@ -152,6 +154,40 @@ func TestSChainWriter(t *testing.T) {
 			},
 			wantError: true,
 		},
+		{
+			description: "Schain in request, host schain defined, source.ext for bidder request should update with appended host schain",
+			giveRequest: openrtb2.BidRequest{
+				Ext:    json.RawMessage(`{"prebid":{"schains":[{"bidders":["testbidder"],"schain":{"complete":1,"nodes":[{"asi":"appnexus.com","sid":"1792","rid":"6a586290-ae54-44e9-8985-d60821f7bb07","hp":1}],"ver":"1.0"}}]}}`),
+				Source: nil,
+			},
+			giveBidder: "testbidder",
+			giveHostSChain: &openrtb_ext.ExtRequestPrebidSChainSChainNode{
+				ASI: "pbshostcompany.com", SID: "00001", RID: "BidRequest", HP: 1,
+			},
+			wantRequest: openrtb2.BidRequest{
+				Ext: json.RawMessage(`{"prebid":{"schains":[{"bidders":["testbidder"],"schain":{"complete":1,"nodes":[{"asi":"appnexus.com","sid":"1792","rid":"6a586290-ae54-44e9-8985-d60821f7bb07","hp":1}],"ver":"1.0"}}]}}`),
+				Source: &openrtb2.Source{
+					Ext: json.RawMessage(`{"schain":{"complete":1,"nodes":[{"asi":"appnexus.com","sid":"1792","rid":"6a586290-ae54-44e9-8985-d60821f7bb07","hp":1},{"asi":"pbshostcompany.com","sid":"00001","rid":"BidRequest","hp":1}],"ver":"1.0"}}`),
+				},
+			},
+		},
+		{
+			description: "No Schain in request, host schain defined, source.ext for bidder request should have just the host schain",
+			giveRequest: openrtb2.BidRequest{
+				Ext:    nil,
+				Source: nil,
+			},
+			giveBidder: "testbidder",
+			giveHostSChain: &openrtb_ext.ExtRequestPrebidSChainSChainNode{
+				ASI: "pbshostcompany.com", SID: "00001", RID: "BidRequest", HP: 1,
+			},
+			wantRequest: openrtb2.BidRequest{
+				Ext: nil,
+				Source: &openrtb2.Source{
+					Ext: json.RawMessage(`{"schain":{"complete":0,"nodes":[{"asi":"pbshostcompany.com","sid":"00001","rid":"BidRequest","hp":1}],"ver":"1.0"}}`),
+				},
+			},
+		},
```

Comment:
Nitpick: to improve readability and continue with the style of this test, I'd suggest defining a couple of schain node constants at the top of this test like so:
```
const hostNode    string = `{"asi":"pbshostcompany.com","sid":"00001","rid":"BidRequest","hp":1}`
const seller1Node string = `{"asi":"directseller1.com","sid":"00001","rid":"BidRequest1","hp":1}`
```
and then using them here so we don't have long inline JSON strings that are hard to read and digest for someone unfamiliar with the logic:
```
@@ -157,7 +160,7 @@ func TestSChainWriter(t *testing.T) {
                {
                        description: "Schain in request, host schain defined, source.ext for bidder request should update with appended host schain",
                        giveRequest: openrtb2.BidRequest{
-                               Ext:    json.RawMessage(`{"prebid":{"schains":[{"bidders":["testbidder"],"schain":{"complete":1,"nodes":[{"asi":"appnexus.com","sid":"1792","rid":"6a586290-ae54-44e9-8985-d60821f7bb07","hp":1}],"ver":"1.0"}}]}}`),
+                               Ext:    json.RawMessage(`{"prebid":{"schains":[{"bidders":["testbidder"],"schain":{"complete":1,"nodes":[` + seller1Node + `],"ver":"1.0"}}]}}`),
                                Source: nil,
                        },
                        giveBidder: "testbidder",
@@ -165,9 +168,9 @@ func TestSChainWriter(t *testing.T) {
                                ASI: "pbshostcompany.com", SID: "00001", RID: "BidRequest", HP: 1,
                        },
                        wantRequest: openrtb2.BidRequest{
-                               Ext: json.RawMessage(`{"prebid":{"schains":[{"bidders":["testbidder"],"schain":{"complete":1,"nodes":[{"asi":"appnexus.com","sid":"1792","rid":"6a586290-ae54-44e9-8985-d60821f7bb07","hp":1}],"ver":"1.0"}}]}}`),
+                               Ext: json.RawMessage(`{"prebid":{"schains":[{"bidders":["testbidder"],"schain":{"complete":1,"nodes":[` + seller1Node + `],"ver":"1.0"}}]}}`),
                                Source: &openrtb2.Source{
-                                       Ext: json.RawMessage(`{"schain":{"complete":1,"nodes":[{"asi":"appnexus.com","sid":"1792","rid":"6a586290-ae54-44e9-8985-d60821f7bb07","hp":1},{"asi":"pbshostcompany.com","sid":"00001","rid":"BidRequest","hp":1}],"ver":"1.0"}}`),
+                                       Ext: json.RawMessage(`{"schain":{"complete":1,"nodes":[` + seller1Node +`,` + hostNode + `],"ver":"1.0"}}`),
                                },
                        },
                },
@@ -184,7 +187,7 @@ func TestSChainWriter(t *testing.T) {
                        wantRequest: openrtb2.BidRequest{
                                Ext: nil,
                                Source: &openrtb2.Source{
-                                       Ext: json.RawMessage(`{"schain":{"complete":0,"nodes":[{"asi":"pbshostcompany.com","sid":"00001","rid":"BidRequest","hp":1}],"ver":"1.0"}}`),
+                                       Ext: json.RawMessage(`{"schain":{"complete":0,"nodes":[` + hostNode + `],"ver":"1.0"}}`),
                                },
                        },
                }
```